### DISCUSSION TITLE
Cyberdeck theme/project

### DISCUSSION THREAD

**Discussion by hicwic** [mathoudebine/turing-smart-screen-python#374] [Themes]
hi,

i made a PR yesterday with some improvement (plot graph to display fps histo), and a demo theme.

Here is an updated version of the theme, you can adapt it even if my PR is not merged. (.ai file provided)

[cyberdeck-theme.zip](https://github.com/mathoudebine/turing-smart-screen-python/files/13161995/cyberdeck-theme.zip)
![preview](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/597dc28e-f2f1-4968-abc5-591668102878)

It's part of a physical project to mod my Cyberpunk themed gaming computer with a fake/realistic Cyberdeck. See screenshot of render bellow.
![cyberdeck](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/98c15ccb-1946-4a6c-a39c-d3d1b917c4cd)

here a photo of the work in progress. I'll print transparent buttons in resin next week
![IMG_20231025_082922095](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/b4c1f57e-2b27-4439-8036-38b7cbf110eb)

You can find the onshape modele here (feel free to copy and modify) : [Cyberdeck 3d Model](https://cad.onshape.com/documents/ec5a8a1a86c62e2a2f1f0047/w/7647896fdbb27cee3fa6da3a/e/ef4e930421864d0da0aaf0df?renderMode=0&uiState=6538b63d7336c21026741c35)

**Comment by 1SayemAhmed** [comment]
bro my fan speed not showing

**Reply by hicwic** [reply]
yes, cause fan speed metric is not implemented on the author repo. 

See my [PR  here](https://github.com/mathoudebine/turing-smart-screen-python/pull/371). You can adapt the sensors_custom to have fan speed.

**Reply by 1SayemAhmed** [reply]
If this load is not intended to sandbox the assembly, please enable the loadFromRemoteSources switch

how to solve this

**Reply by hicwic** [reply]
Don't even know where and why you got this :) maybe if you provide more details of what you're trying to do...

Pretty sure it's not related to my PR.

**Comment by 1SayemAhmed** [comment]
Just copy and replace?

**Reply by hicwic** [reply]
I'll try this evening (cest timezone) to integrate my custom sensors to a dry project.

I'll keep you in touch

**Reply by 1SayemAhmed** [reply]
can you also add cpu or gpu wattage to this theme?

**Reply by hicwic** [reply]
Take this sensors_custom, it will work for fan speed. I removed RTSS library call. (but you need to look the file as the metrics a really dependent of your hardware)

for example this snippet of code take the fan speed from my motherboard, the port named "#2". It'll be great luck if this match your motherboard too :-)
```
            for sensor in sh.Sensors:
                if sensor.SensorType == Hardware.SensorType.Control and "#2" in str(sensor.Name):
                    return float(sensor.Value)
```

this one work for GPU fan speed, but only for Nvidia
```
        gpu = get_hw_and_update(Hardware.HardwareType.GpuNvidia)
        for sensor in gpu.Sensors:
            if sensor.SensorType == Hardware.SensorType.Control:
                return float(sensor.Value)
```

[sensors_custom.zip](https://github.com/mathoudebine/turing-smart-screen-python/files/13179676/sensors_custom.zip)

**Reply by 1SayemAhmed** [reply]
do you have any file for cpu wattage

**Comment by mathoudebine** [comment]
Amazing job @hicwic!
This is the first time I see the smart screen integrated into an object that is not a computer!
Where did you find inspiration for your Cyberdeck? I love it and I look forward to see your progress
Can you give some details about the electronics like buttons and the chip? Will they serve a purpose in the final product or is is just to decorate?
For your PR I did not had the time to review it yet, sorry 😞

**Reply by hicwic** [reply]
Thx,

To be super clear, this is a fake cyberdeck in which the screen is integrated (this cyberdeck will be integrated into my gaming computer), nothing is real, exept the monitoring screen. 
- The chip is an old mp3 decoding module, connected to nothing, i used in a project 20 years ago :), just to make it more "cyperpunk". 
- All the button are not connected to anything. I tested different printing method, and i endend printing them 45° angled, in transparent resin. So they are not transparent but translucid. I put a cheap led strip from "action" just behind so they look backlighted. I painted some of them by behind to react differently to light.
- I put some wire, connected to nothing, just for the "wow effect"

I'll post some pics, maybe next week to show you the progress if interested.

**Comment by mathoudebine** [comment]
I should also add that you did some really nice work on your theme, I love the color palette and everything is perfectly placed to get the most of the small screen, congrats!

**Reply by hicwic** [reply]
thx again, i'm happy with the result. My wife helped a lot to choose colors and font. We did several prototypes before we got this one.

**Comment by hicwic** [comment]
Some updates on the project. Assembly of the first version. I love the result.

Custom cable. Type C + led strip to -> motherboard internal usb
![IMG_20231111_104503735](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/2f20ae6a-3bf7-4330-a1c4-8ac4271a0730)
Assembly
![IMG_20231111_114058400](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/4b1de32a-e397-4174-922d-e7494523cb44)
Installation
![IMG_20231111_120231674](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/add88bdd-0e5d-4359-93f1-69ea753ec3af)
V1 result
![IMG_20231111_122959403_HDR](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/9716c07c-f939-45af-8369-f067f6df4e11)

**Comment by NazariGamer** [comment]
Hi @hicwic thanks for sharing this very nice theme.  I also studied your sensors_custom.py as I want to tweak it because I have AMD hardware and would like to add Fan Speed etc.

I was just wondering where you found documentation on the LibreHardwareMonitorLib.dll file. I had no luck so far.

Many thanks in advance.

**Reply by hicwic** [reply]
Basically I just installed libreharwaremonitor (the exe file) and saw the structure and the name of the sensors reported... Didn't even looked for documentation :-(

**Reply by ooemir23** [reply]
Very good theme but not working amd GPU

**Reply by NazariGamer** [reply]
Hmm was doing the same, but went looking for docs. In my case if I want to capture CPU, Clocks, Bus Speed (see attached image) how would I address that? 

I mean there are spaces in the names (not in yours)  and I thing you get the higher level object (motherboard, nvdiagpu) and iterate over the subitems to get the values?

Many thanks in advance for any guidance you can give me @hicwic 

![image](https://github.com/mathoudebine/turing-smart-screen-python/assets/831364/b752547c-fe23-4db9-bced-519f7a49cba4)

**Reply by hicwic** [reply]
ok, what i do is going to "file => report" and make an export from this graphical view

![report](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/733aadfa-5928-4072-9661-afcd88969971)


then if you open the text file, you will see a "path" near each sensor. From this path you could guess what HardwareType, SubHardwareType, SensorType you need. 

HardwareType are an enum visible on LHM source code. So this is static. For your AMD Gpu it'll be ```GpuAmd```
```
public enum HardwareType
{
    Motherboard,
    SuperIO,
    Cpu,
    Memory,
    GpuNvidia,
    GpuAmd,
    GpuIntel,
    Storage,
    Network,
    Cooler,
    EmbeddedController,
    Psu,
    Battery
}
```


See this example. Hope i'll be clear.
![sensors1](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/d83b0e68-1999-4093-87e7-d3321e110337)

As there is only one subhardware in my motherboard HardwareType, i just iterate over until you find the right sensor. You can select it by name, or by type. Type is directly visible in the path

Here another example with the fan controll speed of my GPU. fan has two type, one in RPM (type "fan"), one in percent (type "control")

So if you look the snippet of code i wrote and the path in LHM export it'easy to understand. As there is only one "control" type sensor in my GpuNvidia, i select it by type. (for the fan speed of my CPU, given by motherboard, there is multiple fan control, so i filtered by type and by name)

![sensors_gpu](https://github.com/mathoudebine/turing-smart-screen-python/assets/1628763/6eef4df5-eb02-44a6-8adb-d05aebdf4374)

Hope this will contribute to your project.

**Reply by mathoudebine** [reply]
You can also run the `external/LibreHardwareMonitor/test_librehardwaremonitor.py`script with Python in Administrator, to get the same information

**Reply by NazariGamer** [reply]
@hicwic @mathoudebine thanks so much. I can now proceed and will share findings later.

**Reply by NazariGamer** [reply]
@hicwic @mathoudebine good evening. I was reading through the code and found some import statements to import modules e.g. library.sensors.sensors
and win32api. However they are not in requirements.txt nor installed on my system. Still the code works.

What is the magic behind this?

**Reply by hicwic** [reply]
From my understanding (not a python dev)

Win32api is from pywin32 in the requirements

library.sensors.sensors is the class sensors in the project

**Comment by StichSince1996** [comment]
Como faço para usar esse tema no formato .data? Tenho a tela de 3.5" da Thurxz (turing).

**Comment by DJRoby19** [comment]
Hi I installed your theme I have a problem that does not display the GPU fan speed in version 3.5.0, instead in version 3.4.0 it worked how can I solve it

**Reply by wsokc** [reply]
Seems like I got the same issues,
9/4/2024 4:06:20 PM [ERROR] Error loading custom sensor class GpuNvidiaFanPercent from sensors_custom.py : 'GpuNvidiaFanPercent' object has no attribute 'last_values'

**Reply by wsokc** [reply]
this what I used for 3.5.0

```
class PHYSMEMORY(CustomDataSource):
    last_val = [math.nan] * 10
    def as_numeric(self) -> float:
        ram = get_hw_and_update(Hardware.HardwareType.Memory)
        for sensor in ram.Sensors:
            if sensor.SensorType == Hardware.SensorType.Data:
                self.value = sensor.Value
                self.last_val.append(self.value)
                self.last_val.pop(0)
                return float(sensor.Value)
            
        logger.error("RAM DATA cannot be read")
        return math.nan
    
    def as_string(self) -> str:
        return f'{int(self.as_numeric())}G'
    
    def last_values(self) -> List[float]:
        return self.last_val
```

**Reply by DJRoby19** [reply]
Did you solve it?
I use the sensors_python.py library with Linux

**Reply by wsokc** [reply]
I used Windows, and I think should be same..
Oh yea, on the custom i add the initiation import from Sensors_libhardwaremonitor
```

import math
import platform

from abc import ABC, abstractmethod
from typing import List

import ctypes
import math
import os
import sys
from statistics import mean
from typing import Tuple

import clr  # Clr is from pythonnet package. Do not install clr package
import psutil
from win32api import *

import library.sensors.sensors as sensors
from library.log import logger

# Import LibreHardwareMonitor dll to Python
lhm_dll = os.getcwd() + '\\external\\LibreHardwareMonitor\\LibreHardwareMonitorLib.dll'
# noinspection PyUnresolvedReferences
clr.AddReference(lhm_dll)
# noinspection PyUnresolvedReferences
clr.AddReference(os.getcwd() + '\\external\\LibreHardwareMonitor\\HidSharp.dll')
# noinspection PyUnresolvedReferences
from LibreHardwareMonitor import Hardware # type: ignore

File_information = GetFileVersionInfo(lhm_dll, "\\")

ms_file_version = File_information['FileVersionMS']
ls_file_version = File_information['FileVersionLS']

logger.debug("Found LibreHardwareMonitorLib %s" % ".".join([str(HIWORD(ms_file_version)), str(LOWORD(ms_file_version)),
                                                            str(HIWORD(ls_file_version)),
                                                            str(LOWORD(ls_file_version))]))

if ctypes.windll.shell32.IsUserAnAdmin() == 0:
    logger.error(
        "Program is not running as administrator. Please run with admin rights or choose another HW_SENSORS option in "
        "config.yaml")
    try:
        sys.exit(0)
    except:
        os._exit(0)

handle = Hardware.Computer()
handle.IsCpuEnabled = True
handle.IsGpuEnabled = True
handle.IsMemoryEnabled = True
handle.IsMotherboardEnabled = True  # For CPU Fan Speed
handle.IsControllerEnabled = True  # For CPU Fan Speed
handle.IsNetworkEnabled = True
handle.IsStorageEnabled = True
handle.IsPsuEnabled = False
handle.Open()
for hardware in handle.Hardware:
    if hardware.HardwareType == Hardware.HardwareType.Cpu:
        logger.info("Found CPU: %s" % hardware.Name)
    elif hardware.HardwareType == Hardware.HardwareType.Memory:
        logger.info("Found Memory: %s" % hardware.Name)
    elif hardware.HardwareType == Hardware.HardwareType.GpuNvidia:
        logger.info("Found Nvidia GPU: %s" % hardware.Name)
    elif hardware.HardwareType == Hardware.HardwareType.GpuAmd:
        logger.info("Found AMD GPU: %s" % hardware.Name)
    elif hardware.HardwareType == Hardware.HardwareType.GpuIntel:
        logger.info("Found Intel GPU: %s" % hardware.Name)
    elif hardware.HardwareType == Hardware.HardwareType.Storage:
        logger.info("Found Storage: %s" % hardware.Name)
    elif hardware.HardwareType == Hardware.HardwareType.Network:
        logger.info("Found Network interface: %s" % hardware.Name)


def get_hw_and_update(hwtype: Hardware.HardwareType, name: str = None) -> Hardware.Hardware:
    for hardware in handle.Hardware:
        if hardware.HardwareType == hwtype:
            if (name and hardware.Name == name) or name is None:
                hardware.Update()
                return hardware
    return None
```

Then followed by our sensor_customs

```
class PHYSMEMORY(CustomDataSource):
    last_val = [math.nan] * 10
    def as_numeric(self) -> float:
        ram = get_hw_and_update(Hardware.HardwareType.Memory)
        for sensor in ram.Sensors:
            if sensor.SensorType == Hardware.SensorType.Data:
                self.value = sensor.Value
                self.last_val.append(self.value)
                self.last_val.pop(0)
                return float(sensor.Value)
            
        logger.error("RAM DATA cannot be read")
        return math.nan
    
    def as_string(self) -> str:
        return f'{int(self.as_numeric())}G'
    
    def last_values(self) -> List[float]:
        return self.last_val
```
In this case, I'm trying to get the phys mem instead of swap mem.
I'm still develop the theme once it completed I will share full pack.

**Comment by DJRoby19** [comment]
Hi do you have any idea to make the amd GPU fan work on linux in the turing version 3.5.0?
in the 3.4.0 version it worked
thanks

**Comment by GPleex** [comment]
Do you have the file to edit the background image? It looks really cool, congratulations!
